[EPM] /package API only lists installable assets. Restore enums.

jfsiii opened the pull request:
## Summary

 * `assets` key of `/package/{pkgkey}` will only list installable Kibana assets (e.g. not `datasets` or `infra-ui-source`?)
 * Restored `enum`s for AssetTypes. Added for some other types, too
 * Found the source of the `"Class extends value undefined is not a constructor or null"` error in Chrome and `"The superclass is not an object."` error in Safari. `common` contained server code, so `hapi` and other server code was ending up in client bundle. Moved hapi code/types to `server/types`

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11482/)
* Commit: eac3628dedd849ec76959f3cf9492e1b82f96537

jfsiii commented on x-pack/legacy/plugins/epm/common/types.ts:
```diff
@@ -61,45 +69,27 @@ export interface AssetParts {
-export type Installed<T = {}> = T & {
-  status: 'installed';
-  savedObject: Installation;
+export type KibanaAssetParts = AssetParts & {
```
Not strictly required, but it narrows the types so `service: 'kibana'` can only have `type: KibanaAssetType`

jfsiii commented on x-pack/legacy/plugins/epm/common/types.ts:
```diff
@@ -61,45 +69,27 @@ export interface AssetParts {
 
-export type NotInstalled<T = {}> = T & {
-  status: 'not_installed';
+export type ElasticsearchAssetParts = AssetParts & {
```
Not strictly required, but it narrows the types so `service: 'elasticsearch'` can only have `type: ElasticsearchAssetType `

jfsiii commented on x-pack/legacy/plugins/epm/common/types.ts:
```diff
@@ -61,45 +69,27 @@ export interface AssetParts {
+
+export type AssetTypeToParts = KibanaAssetTypeToParts & ElasticsearchAssetTypeToParts;
+
+export type AssetsGroupedByServiceByType = Record<
```
Same for these. Not strictly required, but now you cannot enter `elasticsearch` asset types in the `kibana` section

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11494/)
* Commit: 8da2a3f065f859d49fe763fbbc6b1761a4a6f3e4

jfsiii commented:
I moved the server types to a separate file and it ran, but it failed type check because some client code actually used stuff I'd moved to server. API response shapes which includes a Saved Object shape. So I changed the client to import the type from server https://github.com/elastic/kibana/commit/d92df56aa43910b97d3680af18f6b37d0f2c205e

It passed type check and runs, but I wonder if I'm not in the same "it works for now, but..." aside from the logical WTFs of having clients pull from common and server.

/cc @dover

jfsiii commented:
Oh! Maybe I can keep the SO types in common and restrict server/types to the stuff based on hapi or that's otherwise purely server & not suitable for the client bundle

elasticmachine commented:
Pinging @elastic/epm (Feature:EPM)

elasticmachine commented:
Pinging @elastic/logs-metrics-ui (Team:logs-metrics-ui)

jfsiii commented on x-pack/legacy/plugins/epm/public/constants.ts:
```diff
@@ -4,12 +4,12 @@
 export const DisplayedAssets: Record<ServiceName, AssetType[]> = {
-  kibana: ['index-pattern', 'visualization', 'search', 'dashboard'],
-  elasticsearch: ['index-template', 'ingest-pipeline', 'ilm-policy'],
+  kibana: Object.values(KibanaAssetType),
```
We use `Object.values(KibanaAssetType)` in 3 places in this PR. I wasn't sure if I wanted reuse a single exported variable (like `export const kibanaAssetTypes = Object.values(KibanaAssetType)` in, say, `common/constants`) or allow consumers to use the Object however they wish and deal with a little repetition.

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11559/)
* Commit: cac7bd7a59c3915d507d290e502764631c462feb

neptunian commented on x-pack/legacy/plugins/epm/public/constants.ts:
```diff
@@ -6,8 +6,11 @@
+type ServiceNameToAssetTypes = Record<Extract<'kibana', ServiceName>, KibanaAssetType[]> &
+  Record<Extract<'elasticsearch', ServiceName>, ElasticsearchAssetType[]>;
+
+export const DisplayedAssets: ServiceNameToAssetTypes = {
```
Isn't this similar to doing:


```
interface DisplayedAssets{
  kibana: KibanaAssetType[],
  elasticsearch: ElasticsearchAssetType[],
}

```

jfsiii replied:
Yes it's similar. I think the key difference is that an interface doesn't have any type info or restrictions on the key. You could enter `foo` and it would not complain. Also, consumers would only know that the key was a string or number, now they know exactly what the type of the key is.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11590/)
* Commit: 8d2de0b93386a464952f6aaa07c2952e60a16ffa

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/11603/)
* Commit: ccc03473a3419926b17658df3f260f3afbb60a09
